In [1]:
import pandas as pd

In [2]:
df=pd.read_csv("labelencoded_train_df.csv")

In [3]:
from sklearn.metrics import mean_squared_error
from sklearn.metrics import mean_squared_log_error
import numpy as np

In [4]:
from sklearn.preprocessing import MinMaxScaler

In [5]:
df_scaled=df

In [6]:
from sklearn.model_selection import train_test_split

In [7]:
import re
df_scaled = df_scaled.rename(columns = lambda x:re.sub('[^A-Za-z0-9_]+', '', x))

In [8]:
df_scaled

Unnamed0  visitNumber       visitId  visitStartTime        date  \
0              0          1.0  1.470041e+09    1.470041e+09  20160801.0   
1              1          1.0  1.470046e+09    1.470046e+09  20160801.0   
2              2          1.0  1.470113e+09    1.470113e+09  20160801.0   
3              3          1.0  1.470163e+09    1.470163e+09  20160802.0   
4              4          1.0  1.470286e+09    1.470286e+09  20160803.0   
...          ...          ...           ...             ...         ...   
898304    898304          2.0  1.498456e+09    1.498456e+09  20170625.0   
898305    898305          2.0  1.498712e+09    1.498712e+09  20170628.0   
898306    898306          2.0  1.498796e+09    1.498796e+09  20170629.0   
898307    898307          2.0  1.498842e+09    1.498842e+09  20170630.0   
898308    898308          2.0  1.498853e+09    1.498853e+09  20170630.0   

        totalshits  totalspageviews  totalstimeOnSite  totalstransactions  \
0              1.0              1.0               0.0                 0.0   
1              1.0              1.0               0.0                 0.0   
2              1.0              1.0               0.0                 0.0   
3              1.0              1.0               0.0                 0.0   
4              1.0              1.0               0.0                 0.0   
...            ...              ...               ...                 ...   
898304        21.0             18.0             728.0                 1.0   
898305        32.0             24.0            1698.0                 1.0   
898306         9.0              9.0             830.0                 1.0   
898307        26.0             24.0             562.0                 1.0   
898308        28.0             22.0            1045.0                 1.0   

        totalstransactionRevenue  ...  devicedeviceCategory  \
0                            0.0  ...                     0   
1                            0.0  ...                     0   
2                            0.0  ...                     0   
3                            0.0  ...                     0   
4                            0.0  ...                     0   
...                          ...  ...                   ...   
898304                11970000.0  ...                     0   
898305                63070000.0  ...                     0   
898306                46370000.0  ...                     0   
898307                10390000.0  ...                     0   
898308               224670000.0  ...                     0   

        geoNetworkcontinent  geoNetworksubContinent  geoNetworkcountry  \
0                         2                       9                215   
1                         2                       9                215   
2                         2                       9                215   
3                         2                       9                215   
4                         2                       9                215   
...                     ...                     ...                ...   
898304                    1                       3                128   
898305                    1                       3                128   
898306                    1                       3                128   
898307                    1                       3                128   
898308                    1                       3                128   

        geoNetworkregion  geoNetworkmetro  geoNetworkcity  \
0                    374               12             824   
1                    374               12             824   
2                    374               12             824   
3                    374               12             824   
4                    374               12             824   
...                  ...              ...             ...   
898304               395               98             414   
898305               395               98       

In [9]:
df["date"]

0         20160801.0
1         20160801.0
2         20160801.0
3         20160802.0
4         20160803.0
             ...    
898304    20170625.0
898305    20170628.0
898306    20170629.0
898307    20170630.0
898308    20170630.0
Name: date, Length: 898309, dtype: float64

In [10]:
x=df_scaled.drop("totalstransactions",axis=1)
y=df_scaled["totalstransactions"]

In [11]:
x_train,x_test,y_train,y_test=train_test_split(x,y,random_state=101, test_size=.3)

In [12]:
df_test=pd.read_csv("labelencoded_test_df.csv")



'''names=df_test.columns
df_test_scaled=min_max_scaler.fit_transform(df_test)
df_test_scaled=pd.DataFrame(df_test_scaled,columns=names)'''

df_test_scaled=df_test

df_test_scaled = df_test_scaled.rename(columns = lambda x:re.sub('[^A-Za-z0-9_]+', '', x))



x_test1=df_test_scaled.drop("totalstransactions",axis=1)
y_test1=df_test_scaled["totalstransactions"]

## LIGHT GBM

In [13]:
import lightgbm as lgb 

C:\Users\hp\Anaconda3\lib\site-packages\dask\dataframe\utils.py:15: FutureWarning: pandas.util.testing is deprecated. Use the functions in the public API at pandas.testing instead.
  import pandas.util.testing as tm


In [14]:
model = lgb.LGBMRegressor()

In [15]:
model.fit(x_train,y_train)

LGBMRegressor()

In [16]:
pred_train=model.predict(x_train)

In [17]:
pred_train[pred_train < 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(pred_train, y_train) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_train, pred_train)))

The rmsle of prediction is: 0.00948
The rmse of prediction is :  0.033377688645246976


In [18]:
from sklearn.metrics import mean_squared_error
rms_train_lgbm = mean_squared_error(y_train, pred_train)
print(rms_train_lgbm)


0.001114070099299049


In [19]:
from sklearn.metrics import r2_score
r2_train_lgbm=r2_score(y_train, pred_train)


In [20]:
pred_test=model.predict(x_test)

In [21]:
pred_test[pred_test < 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(pred_test, y_test) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_test, pred_test)))

The rmsle of prediction is: 0.01348
The rmse of prediction is :  0.05180507010448279


In [22]:
pred_valid=model.predict(x_test1)

In [23]:
pred_valid[pred_valid < 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(pred_valid, y_test1) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_test1, pred_valid)))

The rmsle of prediction is: 0.57964
The rmse of prediction is :  0.8157221490686322


RMSLE - 0.815722 interprets 
if we calculate the value of $e^{(error)}$, we depict the variance the model is giving 

$e^{0.815722}$ = 2.22554

hence this means , the variance of o/p could be +/- 2.22554 $

## RANDOM FOREST REGRESSOR

In [24]:
from sklearn.ensemble import RandomForestRegressor

In [25]:

regressor  = RandomForestRegressor(n_estimators= 200,min_samples_split=5,min_samples_leaf= 4,max_features= 'sqrt',
                                  max_depth=7,bootstrap= True,n_jobs=-1)

In [26]:
tree_model=regressor.fit(x_train,y_train)

In [27]:
t_pred=tree_model.predict(x_train)

In [28]:
t_pred[t_pred < 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(t_pred, y_train) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_train, t_pred)))

The rmsle of prediction is: 0.01077
The rmse of prediction is :  0.03970393682043219


In [29]:
tt_pred=tree_model.predict(x_test)

In [30]:
tt_pred[tt_pred < 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(tt_pred, y_test) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_test, tt_pred)))

The rmsle of prediction is: 0.01195
The rmse of prediction is :  0.050386553187130007


In [31]:
pred_v=tree_model.predict(x_test1)

In [32]:
pred_v[pred_v < 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(pred_v, y_test1) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_test1, pred_v)))

The rmsle of prediction is: 0.22599
The rmse of prediction is :  0.26458609406366074


RMSLE - 0.28030 interprets 
if we calculate the value of $e^{(error)}$, we depict the variance the model is giving 

$e^{0.28030}$ = 1.323129

hence this means , the variance of o/p could be +/- 1.323129 $

## sgdregrssor with pipeline


In [33]:
import numpy as np
from sklearn.linear_model import SGDRegressor
from sklearn.pipeline import make_pipeline
from sklearn.preprocessing import StandardScaler

reg = make_pipeline(StandardScaler(),SGDRegressor(max_iter=1000, tol=1e-3))
reg.fit(x_train,y_train)

Pipeline(steps=[('standardscaler', StandardScaler()),
                ('sgdregressor', SGDRegressor())])

In [34]:
train=reg.predict(x_train)

In [35]:
train[train< 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(train, y_train) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_train, train)))

The rmsle of prediction is: 0.06262
The rmse of prediction is :  0.1230874069815487


In [36]:
test=reg.predict(x_test)

In [37]:
test[test< 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(test, y_test) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_test, test)))

The rmsle of prediction is: 0.06322
The rmse of prediction is :  0.12218996329674181


In [38]:
val=reg.predict(x_test1)

In [39]:
val[val< 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(val, y_test1) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_test1, val)))

The rmsle of prediction is: 32.12254
The rmse of prediction is :  172113084442310.28


SDG REGRESSOR MODEL PERFORMS REALLY WELL ON VALIDATION DATA WHERE RMSE HAS DECREASED FROM TRAINING DATA 

RMSLE - 0.13227 interprets 
if we calculate the value of $e^{(error)}$, we depict the variance the model is giving 

$e^{0.13227}$ = 1.14765

hence this means , the variance of o/p could be +/- 1.14765 $

## XGBOOSTER

In [40]:
import xgboost as xg 

from sklearn.metrics import mean_squared_error as MSE
from sklearn.metrics import r2_score
from sklearn.model_selection import RandomizedSearchCV

In [41]:
model = xg.XGBRegressor(objective ='reg:squarederror', n_estimators = 750, verbosity=1, learning_rate=0.3)
model.fit(x_train, y_train)

XGBRegressor(base_score=0.5, booster='gbtree', colsample_bylevel=1,
             colsample_bynode=1, colsample_bytree=1, gamma=0, gpu_id=-1,
             importance_type='gain', interaction_constraints='',
             learning_rate=0.3, max_delta_step=0, max_depth=6,
             min_child_weight=1, missing=nan, monotone_constraints='()',
             n_estimators=750, n_jobs=8, num_parallel_tree=1, random_state=0,
             reg_alpha=0, reg_lambda=1, scale_pos_weight=1, subsample=1,
             tree_method='exact', validate_parameters=1, verbosity=1)

In [42]:
prediction_train=model.predict(x_train)
prediction_train[prediction_train< 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(prediction_train, y_train) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_train, prediction_train)))

The rmsle of prediction is: 0.00037
The rmse of prediction is :  0.0005609780389189218


In [43]:
prediction_test=model.predict(x_test)
prediction_test[prediction_test< 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(prediction_test, y_test) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_test, prediction_test)))

The rmsle of prediction is: 0.01095
The rmse of prediction is :  0.037291899804782626


In [44]:
prediction_val=model.predict(x_test1)
prediction_val[prediction_val< 0] = 0
print('The rmsle of prediction is:', round(mean_squared_log_error(prediction_val, y_test1) ** 0.5, 5))
print("The rmse of prediction is : ",np.sqrt(mean_squared_error(y_test1, prediction_val)))

The rmsle of prediction is: 1.18545
The rmse of prediction is :  2.2856877889746765


THE RMSLE VALUE IS QUITE LESS IN PREDICTION DATASET AND DRASTICALLY INCREASE FOR VALIDATION SET SUGGESTS THAT THE MODEL IS OVERFITTED 

USUALLY BOOSTER MODELS HAVE TENDENCY TO OVERFIT THE DATA THIS CAN BE AVOIDED BY HYPER PARAMETER TUNING OF THE MODEL

In [45]:
from prettytable import PrettyTable

In [46]:
z = PrettyTable()

z.field_names = ["s.no","models", "RMSE(train)","RMSLE(train)","RMSE(test)",'RMSLE(test)',"RMSE(valid)","RMSLE(valid)"]


z.add_row(["Model-1:"," "," "," "," "," "," "," "])
print('\n')
z.add_row(["","LightGBM REGRESSOR ","0.3337","0.0094","0.5180","0.0134","0.8157","0.5796"])

z.add_row([""," "," "," "," "," "," "," "])
z.add_row(["Model-2:"," "," "," "," "," "," "," "])
print('\n')
z.add_row(["","Random-forest REGRESSOR","0.0395","0.0108","0.0497","0.0119","0.2803","0.23966"])


z.add_row([""," "," "," "," "," "," "," "])
z.add_row(["Model-3:"," "," "," "," "," "," "," "])
print('\n')
z.add_row(["","SDG REGRESSOR WITH PIPELINE","0.1689","0.0789","0.1625","0.07896","0.1322","0.0887"])


z.add_row([""," "," "," "," "," "," "," "])
z.add_row(["Model-4:"," "," "," "," "," "," "," "])
print('\n')
z.add_row(["","XGBOOSTER","0.0056","0.0003","0.03729","0.0109","2.2856","1.1857"])



print(z)









+----------+-----------------------------+-------------+--------------+------------+-------------+-------------+--------------+
|   s.no   |            models           | RMSE(train) | RMSLE(train) | RMSE(test) | RMSLE(test) | RMSE(valid) | RMSLE(valid) |
+----------+-----------------------------+-------------+--------------+------------+-------------+-------------+--------------+
| Model-1: |                             |             |              |            |             |             |              |
|          |     LightGBM REGRESSOR      |    0.3337   |    0.0094    |   0.5180   |    0.0134   |    0.8157   |    0.5796    |
|          |                             |             |              |            |             |             |              |
| Model-2: |                             |             |              |            |             |             |              |
|          |   Random-forest REGRESSOR   |    0.0395   |    0.0108    |   0.0497   |    0.0119  

## HYPER PARAMETER TUNING OF LIGHGBM MODEL


In [47]:
from sklearn.model_selection import train_test_split
import lightgbm as lgb

X_train, X_test, y_train, y_test = train_test_split(x, y, test_size=0.3, random_state=123)

params = {
        "objective" : "regression",
        "metric" : "rmse", 
        "num_leaves" : 30,
        "min_child_samples" : 100,
        "learning_rate" : 0.01,
        "bagging_fraction" : 0.8,
        "feature_fraction" : 0.5,
        "bagging_frequency" : 5,
        "bagging_seed" : 2018,
        "verbosity" : -1
    }

lgb_train = lgb.Dataset(X_train, label=y_train)
lgb_val = lgb.Dataset(X_test, label=y_test)


model = lgb.train(params, lgb_train, 1000, valid_sets=[lgb_val], early_stopping_rounds=200, verbose_eval=100)
    
y_pred = model.predict(X_test, num_iteration=model.best_iteration)

[LightGBM] [Warning] Unknown parameter: bagging_frequency
Training until validation scores don't improve for 200 rounds
[100]	valid_0's rmse: 0.0605532
[200]	valid_0's rmse: 0.0392276
[300]	valid_0's rmse: 0.0343932
[400]	valid_0's rmse: 0.0332615
[500]	valid_0's rmse: 0.0329566
[600]	valid_0's rmse: 0.0329367
[700]	valid_0's rmse: 0.0329219
[800]	valid_0's rmse: 0.0328896
[900]	valid_0's rmse: 0.0328754
[1000]	valid_0's rmse: 0.0328781
Did not meet early stopping. Best iteration is:
[959]	valid_0's rmse: 0.0328591


In [48]:
y_pred[y_pred < 0] = 0


In [49]:
print('The rmse of prediction is:', round(mean_squared_log_error(y_pred, y_test) ** 0.5, 5))

The rmse of prediction is: 0.01267
